src/Libraries/RevitNodes/Elements/DirectShape.cs

```diff
@@ -0,0 +1,471 @@
+        /// Sets the internalDirectShape's shape to point to some geometry,
+        /// this method also generates tessellated geometry from the protogeometry object
+        /// and sets the material of the generated Revit faces
+        /// </summary>
+        /// <param name="shapeReference"></param>
+        /// <param name="materialId"></param>
+        private void InteralSetShape(DesignScriptEntity shapeReference, ElementId materialId,string currentSyncId)
+        {
+            //if the elementID for the current directShape revitElement exists on the input Geometry AND
+            //the value stored at that key is equal to the materialId we're trying to set AND
+            //the previousState's syncId (in the tags dictionary) equals the current syncId (from trace)
+            //then we do not need to regenerate
+
+            //first lookup the state on the input geometry
+            var previousState = shapeReference.Tags.LookupTag(this.InternalElementId.ToString()) as DirectShapeState;
+            //then compare values
+            if (previousState != null && previousState.materialId == materialId.IntegerValue && previousState.syncId == currentSyncId)
+            {
+                return;
+            }
+
+            TransactionManager.Instance.EnsureInTransaction(Document);
+
+            var tessellatedShape = GenerateTessellatedGeo(shapeReference, materialId);
+
+            this.InternalDirectShape.SetShape(tessellatedShape);
+
+            //update the value in trace, since we had to modify the geometry we need a new syncId
+            var updatedTraceData = new DirectShapeState(this, Guid.NewGuid().ToString(), materialId);
+            ElementBinder.SetRawDataForTrace(updatedTraceData);
+            //place new values in the tags dict
+            if (shapeReference.Tags.LookupTag(this.InternalElementId.ToString())== null)
+            {
+                shapeReference.Tags.AddTag(this.InternalElementId.ToString(),updatedTraceData );
+            }
+            else
+            {
+                var storedState = shapeReference.Tags.LookupTag(this.InternalElementId.ToString()) as DirectShapeState;
+                storedState.syncId = updatedTraceData.syncId;
+                storedState.materialId = updatedTraceData.materialId;
+                storedState.IntID = updatedTraceData.IntID;
+                storedState.StringID = updatedTraceData.StringID;
+            }
+
+            TransactionManager.Instance.TransactionTaskDone();
+        }
+
+        /// <summary>
+        /// Sets the internalDirectShape to have a new name
+        /// if this method detects the default string it generates a more specific name
+        /// </summary>
+        /// <param name="name"></param>
+        private void InteralSetName(DesignScriptEntity shapeReference,string name)
+        {
+            if (name == DEFAULT_NAME)
+            {
+                name = name.Replace("...", shapeReference.ToString());
```

Comment:
Use String.Format and follow the ProtoGeometry style, as shown here:

https://github.com/DynamoDS/Dynamo/blob/master/src/Libraries/Display/Display.cs#L144